ARCs/arc-0014.md

```diff
@@ -0,0 +1,247 @@
+
+> ARC-14 does not define any standard for session handling
+
+In a blockchain context, users do not have credentials. Conversely, in the Web3 users are identified through their unique blockchain public addresses (accounts for Algorand). Therefore, traditional authentication mechanisms result impractical. To interact with a dApp, users connect their wallet to the application. Using an authentication layer based on credentials on top of that mechanism might result redundant and inefficient.
+
+In Web3, dApps and traditional systems will be increasingly more interconnected. It is not difficult to imagine users consuming services both from a dApp and a traditional system simultaneously. In that case, a user should be authenticated first through traditional SSOs, and then connecting their blockchain wallet. A better approach should enforce a single SSO procedure.
+
+This ARC provides the standard to authenticate users leveraging their Algorand accounts, and without relying on credentials verification.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>
+> Comments like this are non-normative.
+
+Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects, and all JSON string types are UTF-8 encoded.
+
+This ARC uses interchangeably the terms "*blockchain address*", "*public key*", and "*PK*" to indicate the on-chain address of a blockchain identity, and in particular of an Algorand account.
+
+### Overview
+
+This document describes a standard approach to authenticate Users with their blockchain identity. In particular, an Algorand address is used as a *unique identifier*, and the secret key of that address is used to digitally sign a *message* as a proof of identity for a  Verifier.
+
+To sum up, given an Algorand account <*PKa, SKa*>, this ARC defines the standards for:
+
+- create an [ARC-14](./arc-0014.md) compliant digital signature with *SKa*;
+- verify an ARC-14 compliant digital signature with *PKa*.
+
+### Assumptions
+
+The standard proposed in this document works under the following assumptions:
+
+- User and Verifier communicates over secure SSL/TLS encrypted channels;
+- The Verifier knows the Users’ *PKa*;
+- For each *PKa* the Verifier generates a unique message to be signed;
+- The message changes for each authentication request - avoid replay attack;
+- Users are the only custodians of the *SKa*;
+- The secret keys are stored into a wallet;
+- Users won’t use different *PKa* for authentication;
+- Users won’t change their public address for authentication;
+- Users **MUST** use Algorand compliant keys to sign the messages;
+- Multisig and LogicSig are not supported;
+- Users **MUST NOT** use rekeyed accounts.
```

Comment:
I think this is a mistake.

Rekeying is something that is entirely in the PK owners control.  To rekey, a signed transaction must first be issued.  This is approval by that PK to rekey to a new SK - delegating that SK.  If this proposal accepts signing via the PK 'at all' - then it can't selectively decide that the account using that same SK to rekey is not acceptable.  If you can't trust the owner for action B, why are you trusting it for anything else.  Authentication should be "account based", not PK-based.

Authentication is for an *account* and I, as an account holder, control that account via its PK/SK.  If I as that 'owner' need to rekey for various reasons (security for example - perhaps because of a suspected compromise), then I can simply rekey my account to a new auth PK/SK and continue to use my account.

Not allowing rekeying blocks something fundamental, and useful in the protocol.

Reply:
Can we use the auth address of the account instead of the PK?

Reply:
Hi @pbennett thank you.

The initial purpose of this ARC is to authenticate a `User` to a `Verifier` (e.g. a backend) with his Algorand Account <PK,SK>, replacing traditional (username, password) authentication.

In the case of rekey, the direct link between a `User` and the Algorand Account could not persist anymore -As a `Verifier`, I have no Idea whether `User A` has rekeyed to a new SK he still owns. `User A`, indeed, could rekey to `User B`'s SK. In that case the `Verifier` should not trust `User B` to authenticate with `User A`'s PK.

@SudoWeezy What do you mean with Account's auth address?

Reply:
Forgot what I said. It will be part of an ARC-extension to handle the rekeyed account. We can look for auth-address authentication instead of public key authentication.

Reply:
The Account can be associated with a server during Registration by signing a transaction to register a new PK specific to that server. This would allow WebAuthn for login, so authentication would not need to involve the wallet, and would fit in with passwordless login trends.

That PK itself could expire in time and whenever the verifier is online wrt to the Algorand network and detects a rekey auth address change.

Reply:
> The initial purpose of this ARC is to authenticate a `User` to a `Verifier` (e.g. a backend) with his Algorand Account <PK,SK>, replacing traditional (username, password) authentication.

The 'user' here should be the account - controlled via their keys.  What represents and signs for that user should be tied to the account - and in all ways they control that.  It can't be rekeyed without THEM rekeying it.  Blocking that is blocking that user.  Accounts should be able to be permanent and rekeying is partly what allows them to be so.  I strongly disagree with the binding of single PK/SK as 'user.'

> In the case of rekey, the direct link between a `User` and the Algorand Account could not persist anymore -As a `Verifier`, I have no Idea whether `User A` has rekeyed to a new SK he still owns. `User A`, indeed, could rekey to `User B`'s SK. In that case the `Verifier` should not trust `User B` to authenticate with `User A`'s PK.

It doesn't matter - that's literally none of your concern or business.  How can you verify they didn't hand off their SK to another party for your PK/SK bound process?  You can't - you either trust the owner of the SK or you don't.  You trust that SK - so you need to trust that if that SK rekeyed to a different auth'ing SK - the owner of that PK/SK did it for a reason.

Reply:
@pbennett > I strongly disagree with the binding of single PK/SK as 'user.' 

Unless the PK/SK is a proxy for the account created during a registration step.

Reply:
That should always just be the account (a PK).
That account can be signed by its SK, or by another via rekeying and it having a different auth address.
Anything wanting to 'store' anything should just store the account PK.

Reply:
> That should always just be the account (a PK). That account can be signed by its SK, or by another via rekeying and it having a different auth address. Anything wanting to 'store' anything should just store the account PK.

Not necessarily. A registration step may associate an Account with a [temporary] PK specific to that server. This permits scenarios where a multisig for example can authenticate registration for a delegated user at a service, etc

Reply:
This arc isn't for 'registration' though - if a service wants to associate additional details with an account or particular PK, that's up to it how to handle this and I think discussing it here is sidetracking the issue.

I think most importantly, if I have an account, MY account - and can't "authenticate" with that - then I think this proposal fails to meet its stated requirement.

Reply:
> This arc isn't for 'registration' though - if a service wants to associate additional details with an account or particular PK, that's up to it how to handle this and I think discussing it here is sidetracking the issue.
> 
> I think most importantly, if I have an account, MY account - and can't "authenticate" with that - then I think this proposal fails to meet its stated requirement.

Well I certainly cannot agree with that. Unless the system is simply presenting public information, like Account NFTs, where authentication would not even be necessary, for most conceivable scenarios authentication must be preceded by registration. In that case user experiences and security could likely be improved by using the registration step to generate a PK so that wallets would not need to be involved in the process. In any case this ARC *should* include registration with verifier, not least to address GDPR when the account is stored on server

Reply:
An account by itself is pseudonymous.  A site choosing to attach, store, and track additional information related to that is its own business and if it has GDPR issues because of *that specific choice* that's independent from this ARC.

Again - completely independent of saying this account signed a transaction.  If a particular service demands some kind of KYC/etc to link a single identity with an account (apparently what you want does) - that's independent.
Authenticating an account is about proving the client can sign with those keys 'now' - and isn't about proving they're "user X" as your system defines that - that's independent.  Authentication is not Authorization, nor is it unique "Identity."

Sites that want to authenticate solely based on blockchain identities, where an account IS the identifying information - then just having a 'connected' wallet (or something that fakes it and just sends an address to an api with nothing else) that basically says 'trust me bro - here's my account list' isn't sufficient.  If the app only ever presents transactions to sign/submit and has no overt bot issues then it isn't necessary for a distinct auth step, but for apps that want to verify the connected account actually controls the specified account and then perform traditional session management to gate bot mechanisms - an authentication/signing mechanism is completely necessary.

Reply:
An account by itself is pseudonymous AND personal information and must be stored on the server for any kind of useful account related session. An absent explicit registration step means implicit registration, the account stored at least for session resumption and for any other functionality. While yes this arc is about defining a message that should at some point be signed to prove account control, I don't think that is useful enough without the context of an overall protocol.

Reply:
Session management is completely independent and is separate from this proposal.

Reply:
Omg! Really?? I must have missed that somehow. Thanks for the heads up.